File: nodejs-app/src/ingest.ts
import { ChromaClient, OpenAIEmbeddingFunction } from 'chromadb';
import OpenAI from 'openai';
import { readFileSync } from 'fs';
import { glob } from 'glob';
import path from 'path';
import { createHash } from 'crypto';

interface Document {
  content: string;
  filename: string;
  filepath: string;
}

interface ChunkMetadata {
  filename: string;
  filepath: string;
  chunk_index: number;
  total_chunks: number;
  [key: string]: any;
}

function getChromaClient(): ChromaClient {
  const chromaHost = process.env.CHROMA_HOST;
  const chromaPort = parseInt(process.env.CHROMA_PORT || '8000');
  const chromaSsl = process.env.CHROMA_SSL?.toLowerCase() === 'true';
  
  if (chromaHost) {
    const auth = process.env.CHROMA_AUTH_TOKEN 
      ? { provider: 'token', credentials: process.env.CHROMA_AUTH_TOKEN }
      : undefined;
    
    return new ChromaClient({
      path: `http${chromaSsl ? 's' : ''}://${chromaHost}:${chromaPort}`,
      auth
    });
  } else {
    return new ChromaClient();
  }
}

function readMarkdownFiles(dataDir: string = 'data'): Document[] {
  const mdFiles = glob.sync(path.join(dataDir, '*.md'));
  const documents: Document[] = [];
  
  for (const filePath of mdFiles) {
    const content = readFileSync(filePath, 'utf-8');
    documents.push({
      content,
      filename: path.basename(filePath),
      filepath: filePath
    });
  }
  
  return documents;
}

function chunkText(text: string, chunkSize: number = 1000, overlap: number = 200): string[] {
  const words = text.split(' ');
  const chunks: string[] = [];
  
  for (let i = 0; i < words.length; i += chunkSize - overlap) {
    const chunk = words.slice(i, i + chunkSize).join(' ');
    if (chunk.trim()) {
      chunks.push(chunk);
    }
  }
  
  return chunks;
}

async function getEmbeddings(texts: string[], openaiClient: OpenAI): Promise<number[][]> {
  const response = await openaiClient.embeddings.create({
    model: 'text-embedding-3-small',
    input: texts
  });
  
  return response.data.map(embedding => embedding.embedding);
}

function createDocumentId(content: string, filename: string, chunkIdx: number): string {
  const contentHash = createHash('md5').update(content).digest('hex').substring(0, 8);
  return `${filename}_${chunkIdx}_${contentHash}`;
}

export async function ingestDocuments(dataDir: string = 'data', collectionName: string = 'nodejs-app'): Promise<void> {
  // Initialize clients
  const openaiClient = new OpenAI({
    apiKey: process.env.OPENAI_API_KEY
  });
  const chromaClient = getChromaClient();
  
  const embedder = new OpenAIEmbeddingFunction({
    openai_api_key: process.env.OPENAI_API_KEY!,
    openai_model: 'text-embedding-3-small'
  });
  
  // Get or create collection and clear existing data
  let collection;
  try {
    collection = await chromaClient.getCollection({ 
      name: collectionName,
      embeddingFunction: embedder
    });
    console.log(`Found existing collection: ${collectionName}`);
    // Clear all existing data
    await collection.delete();
    console.log(`Cleared existing data from collection: ${collectionName}`);
  } catch (error) {
    // Collection doesn't exist, create it
    collection = await chromaClient.createCollection({ 
      name: collectionName,
      embeddingFunction: embedder
    });
    console.log(`Created new collection: ${collectionName}`);
  }
  
  // Read markdown files
  const documents = readMarkdownFiles(dataDir);
  console.log(`Found ${documents.length} markdown files`);
  
  // Process each document
  const allChunks: string[] = [];
  const allMetadatas: ChunkMetadata[] = [];
  const allIds: string[] = [];
  
  for (const doc of documents) {
    console.log(`Processing: ${doc.filename}`);
    
    // Chunk the document
    const chunks = chunkText(doc.content);
    console.log(`  Created ${chunks.length} chunks`);
    
    // Create metadata for each chunk
    for (let i = 0; i < chunks.length; i++) {
      const chunk = chunks[i];
      allChunks.push(chunk);
      allMetadatas.push({
        filename: doc.filename,
        filepath: doc.filepath,
        chunk_index: i,
        total_chunks: chunks.length
      });
      allIds.push(createDocumentId(chunk, doc.filename, i));
    }
  }
  
  console.log(`Total chunks to process: ${allChunks.length}`);
  
  // Add to ChromaDB (ChromaDB will handle embeddings automatically with the embedding function)
  console.log('Storing in ChromaDB...');
  await collection.add({
    documents: allChunks,
    metadatas: allMetadatas,
    ids: allIds
  });
  
  console.log(`Successfully ingested ${allChunks.length} chunks into collection '${collectionName}'`);
}

// Run if called directly
if (import.meta.url === `file://${process.argv[1]}`) {
  ingestDocuments().catch(console.error);
}

File: python-app/ingest.py
import os
import glob
import chromadb
from openai import OpenAI
from typing import List, Dict
import hashlib


def read_markdown_files(data_dir: str = "data") -> List[Dict[str, str]]:
    """Read all markdown files from the data directory."""
    md_files = glob.glob(os.path.join(data_dir, "*.md"))
    documents = []
    
    for file_path in md_files:
        with open(file_path, 'r', encoding='utf-8') as file:
            content = file.read()
            documents.append({
                "content": content,
                "filename": os.path.basename(file_path),
                "filepath": file_path
            })
    
    return documents


def chunk_text(text: str, chunk_size: int = 1000, overlap: int = 200) -> List[str]:
    """Split text into overlapping chunks."""
    words = text.split()
    chunks = []
    
    for i in range(0, len(words), chunk_size - overlap):
        chunk = " ".join(words[i:i + chunk_size])
        if chunk.strip():
            chunks.append(chunk)
    
    return chunks


def get_embeddings(texts: List[str], openai_client: OpenAI) -> List[List[float]]:
    """Get embeddings for a list of texts using OpenAI."""
    response = openai_client.embeddings.create(
        model="text-embedding-3-small",
        input=texts
    )
    return [embedding.embedding for embedding in response.data]


def create_document_id(content: str, filename: str, chunk_idx: int) -> str:
    """Create a unique ID for a document chunk."""
    content_hash = hashlib.md5(content.encode()).hexdigest()[:8]
    return f"{filename}_{chunk_idx}_{content_hash}"


def get_chroma_client():
    """Create ChromaDB client based on environment variables."""
    # Check for remote ChromaDB configuration
    chroma_host = os.getenv("CHROMA_HOST")
    chroma_port = os.getenv("CHROMA_PORT", "8000")
    chroma_ssl = os.getenv("CHROMA_SSL", "false").lower() == "true"
    chroma_headers = {}
    
    # Optional authentication
    if os.getenv("CHROMA_AUTH_TOKEN"):
        chroma_headers["Authorization"] = f"Bearer {os.getenv('CHROMA_AUTH_TOKEN')}"
    
    if chroma_host:
        # Remote ChromaDB instance
        return chromadb.HttpClient(
            host=chroma_host,
            port=int(chroma_port),
            ssl=chroma_ssl,
            headers=chroma_headers
        )
    else:
        # Local ChromaDB instance
        return chromadb.Client()


def ingest_documents(data_dir: str = "data", collection_name: str = "python-app"):
    """Main ingestion function."""
    # Initialize clients
    openai_client = OpenAI()  # Expects OPENAI_API_KEY environment variable
    chroma_client = get_chroma_client()
    
    # Get or create collection and clear existing data
    try:
        collection = chroma_client.get_collection(name=collection_name)
        print(f"Found existing collection: {collection_name}")
        # Clear all existing data
        collection.delete()
        print(f"Cleared existing data from collection: {collection_name}")
    except Exception:
        # Collection doesn't exist, create it
        collection = chroma_client.create_collection(name=collection_name)
        print(f"Created new collection: {collection_name}")
    
    # Read markdown files
    documents = read_markdown_files(data_dir)
    print(f"Found {len(documents)} markdown files")
    
    # Process each document
    all_chunks = []
    all_metadatas = []
    all_ids = []
    
    for doc in documents:
        print(f"Processing: {doc['filename']}")
        
        # Chunk the document
        chunks = chunk_text(doc['content'])
        print(f"  Created {len(chunks)} chunks")
        
        # Create metadata for each chunk
        for i, chunk in enumerate(chunks):
            all_chunks.append(chunk)
            all_metadatas.append({
                "filename": doc['filename'],
                "filepath": doc['filepath'],
                "chunk_index": i,
                "total_chunks": len(chunks)
            })
            all_ids.append(create_document_id(chunk, doc['filename'], i))
    
    print(f"Total chunks to process: {len(all_chunks)}")
    
    # Get embeddings in batches to avoid API limits
    batch_size = 100
    all_embeddings = []
    
    for i in range(0, len(all_chunks), batch_size):
        batch_chunks = all_chunks[i:i + batch_size]
        print(f"Getting embeddings for batch {i//batch_size + 1}/{(len(all_chunks)-1)//batch_size + 1}")
        
        batch_embeddings = get_embeddings(batch_chunks, openai_client)
        all_embeddings.extend(batch_embeddings)
    
    # Add to ChromaDB
    print("Storing in ChromaDB...")
    collection.add(
        documents=all_chunks,
        metadatas=all_metadatas,
        ids=all_ids,
        embeddings=all_embeddings
    )
    
    print(f"Successfully ingested {len(all_chunks)} chunks into collection '{collection_name}'")
    return collection


if __name__ == "__main__":
    ingest_documents()